content/tutorials/numpy_integration/grass_numpy_integration.qmd

```diff
@@ -0,0 +1,253 @@
+This way GRASS provides the array as the result of the call:
+
+```{python}
+import numpy as np
+
+fractal = tools.r_surf_fractal(output=np.array, seed=6)
+```
+
+::: {.callout-note title="NumPy arrays in GRASS version < 8.5"}
+
+Directly passing NumPy arrays to GRASS tools and receiving them back is a new feature in GRASS v8.5.
+If you work with older versions of GRASS, you can use
+[grass.script.array](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array) and [grass.script.array3d](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array3d):
+
+```{python}
+import grass.script as gs
+import grass.script.array as ga
+
+gs.run_command("r.surf.fractal", seed=6, output="fractal")
+fractal = ga.array("fractal")
+```
+
+:::
+
+Now we can display `fractal` array e.g., using matplotlib library:
+
+```{python}
+import matplotlib.pyplot as plt
+
+plt.figure()
+plt.imshow(fractal, cmap='terrain')
+plt.colorbar(label='Elevation')
+plt.title('Topography from r.surf.fractal')
+plt.xlabel('X-coordinate')
+plt.ylabel('Y-coordinate')
+plt.show()
+```
+
+![Fractal surface generated with r.surf.fractal displayed using matplotlib](fractal_numpy.webp)
+
+We can modify the array:
+
+```{python}
+fractal *= 0.1
+fractal = np.abs(fractal)
+
+plt.figure()
+plt.imshow(fractal, cmap='terrain')
+plt.colorbar(label='Elevation')
+plt.title('Modified topography from r.surf.fractal')
+plt.xlabel('X-coordinate')
+plt.ylabel('Y-coordinate')
+plt.show()
+```
+
+![Modified fractal surface](fractal_numpy_abs.webp)
+
+# From GRASS to Landlab
+
+Now, let's use Landlab's modeling capabilities to burn in an initial drainage network using the
+Landlab's [Fastscape Eroder](https://landlab.readthedocs.io/en/latest/generated/api/landlab.components.stream_power.fastscape_stream_power.html).
+
+Create the `RasterModelGrid` specifying the same dimensions we used for creating the fractal surface.
+Add the terrain elevations as a 1D array (flattened with `ravel`) to the grid's nodes under the standard field name "topographic__elevation".
+
+```{python}
+from landlab import RasterModelGrid
+
+grid = RasterModelGrid((rows, cols), xy_spacing=resolution)
+grid.add_field("topographic__elevation", fractal.ravel(), at="node")
+```
+
+Run the erosion of the landscape, extract the resulting NumPy array and visualize it:
+
+```{python}
+from landlab.components import LinearDiffuser
+from landlab.components import FlowAccumulator, FastscapeEroder
+
+fa = FlowAccumulator(grid, flow_director="D8")
+fsp = FastscapeEroder(grid, K_sp=0.01, m_sp=0.5, n_sp=1.0)
+ld = LinearDiffuser(grid, linear_diffusivity=1)
+
+for i in range(100):
+    fa.run_one_step()
+    fsp.run_one_step(dt=1.0)
+    ld.run_one_step(dt=1.0)
+
+elevation = grid.at_node['topographic__elevation'].reshape(grid.shape)
```

Comment:
Perhaps break the code block here and add text to emphasize another conversion happening.